lib/spack/llnl/util/tty/log.py

```diff
@@ -89,42 +130,127 @@ def __init__(self, stream):
+        # retore input settings on terminal when backgrounded, so we
+        # don't leave the user's terminal in non-canonical, non-echo mode
+        self._restore_input()
+
+        # Run the default SIGTSTP handler for this process to actually stop it.
+        # We reinstall _tstp_handler on SIGCONT.
+        signal.signal(signal.SIGTSTP, self.old_tstp_handler)
+        os.kill(os.getpid(), signal.SIGTSTP)
+
+    def _cont_handler(self, signum, frame):
+        """Handler to restore or enable term settings on SIGCONT."""
+        # SICONT happens when started both in the foreground and background.
+        if self._is_background():
+            # restore defaults if we find ourselves in the background. This
+            # will likely have been done already by _tstp_handler, but this is
+            # *just in case* we get SIGSTOP/SIGCONT (SIGSTOP can't be caught)
+            self._restore_input()
+        else:
+            # re-enable keyboard input if we're newly running in the foreground
+            self._enable_keyboard_input()
+
+        # reinstall the SIGTSTP handler to disable when resuming execution
+        signal.signal(signal.SIGTSTP, self._tstp_handler)
+
     def __enter__(self):
-        """Enable immediate keypress input on stream.
+        """Enable immediate keypress input, while this process is foreground.
 
         If the stream is not a TTY or the system doesn't support termios,
         do nothing.
         """
         self.old_cfg = None
+        self.old_tstp_handler = None
+        self.old_cont_handler = None
 
         # Ignore all this if the input stream is not a tty.
         if not self.stream or not self.stream.isatty():
-            return
+            return self
 
         # If this fails, self.old_cfg will remain None
-        if termios and not _is_background_tty():
-            # save old termios settings
-            old_cfg = termios.tcgetattr(self.stream)
+        if termios:
+            # save old termios settings and signal handlers to restore later
+            self.old_cfg = termios.tcgetattr(self.stream)
+            self.old_tstp_handler = signal.getsignal(signal.SIGTSTP)
+            self.old_cont_handler = signal.getsignal(signal.SIGCONT)
 
-            try:
-                # create new settings with canonical input and echo
-                # disabled, so keypresses are immediate & don't echo.
-                self.new_cfg = termios.tcgetattr(self.stream)
-                self.new_cfg[3] &= ~termios.ICANON
-                self.new_cfg[3] &= ~termios.ECHO
+            # if we exit abnormally, make sure terminal is restored
+            atexit.register(self._restore_input)
 
-                # Apply new settings for terminal
-                termios.tcsetattr(self.stream, termios.TCSADRAIN, self.new_cfg)
-                self.old_cfg = old_cfg
+            # add handlers to disable/enable keyboard input when process
+            # moves from foreground to background.
```

Comment:
I'd replace "from foreground to background" with "between foreground and background"

Reply:
done